### PR TITLE
various updates

### PR DESCRIPTION
**Pull request by joshday**
- julia0.6-pre -> julia0.6
- immutable/type -> struct/mutable struct
- Cleaned up docstrings
- Changed `learn!` to `update!` inside the main loop.  This makes a distinction:
  - `update!` implements a single iteration
  - `learn!` runs at least one iteration according to MetaLearner
- Made MetaLearner immutable (should it be mutable?)
- Ability to send data to a MetaLearner in an online or offline fashion.  Ex:
```julia
learn!(model, meta, data, LearnType.Offline())
```

### CODE REVIEW

File: src/strategies.jl [hunk 1]
```diff
@@ -69,10 +75,13 @@ function finished(strat::Converged, model, i)
...
 
-# -------------------------------------------------------------
-
-"Finished when `‖f(model) - goal‖ ≦ tol`"
+#-----------------------------------------------------------------------# ConvergedTo
+"""
+    ConvergedTo(f, goal; tol=1e-6, every=1)
+Stop learning when `‖f(model) - goal‖ ≦ tol`.
+"""
 type ConvergedTo{V} <: LearningStrategy
```
Comment (Evizero) [inline src/strategies.jl]: `struct`?

File: src/strategies.jl [hunk 2]
```diff
@@ -92,10 +101,11 @@ function finished(strat::ConvergedTo, model, i)
...
-# -------------------------------------------------------------
-
-
-"A sub-strategy to do something each iteration."
+#-----------------------------------------------------------------------# IterFunction
+"""
+    IterFunction(f, b=1)
+Call `f(model, i)` every `b` iterations.
+"""
 immutable IterFunction <: LearningStrategy
```
Comment (Evizero) [inline src/strategies.jl]: also `struct`

File: src/strategies.jl [hunk 3]
```diff
@@ -105,12 +115,13 @@ function iter_hook(strat::IterFunction, model, i)
     if mod1(i, strat.every) == strat.every
         strat.f(model, i)
     end
-    return
```
Comment (Evizero) [inline src/strategies.jl]: will this change not introduce a bit of a type instability? With `return` its clear that the return type is `Void`, now its `Union{Void,WhateverStratDotFReturns}`

**Comment by Evizero** [comment]
I like this change a lot

File: src/strategies.jl (hunk 3 above)
Comment (joshday) [reply src/strategies.jl]: Yep, good call.

Comment (joshday) [reply src/strategies.jl]: I'll add some `@inferred` tests before merging

File: src/metalearner.jl [hunk 4]
```diff
@@ -1,34 +1,54 @@
...
         iter_hook(meta, model, i)
         finished(meta, model, i) && break
+        i += 1
     end
     post_hook(meta, model)
 end
 
+
 # return nothing forever
 type InfiniteNothing end
```
Comment (Evizero) [inline src/metalearner.jl]: also a sneaky hidden `struct`

File: src/metalearner.jl [hunk 5]
```diff
@@ -71,8 +89,3 @@ function make_learner(args...; kw...)
     end
     MetaLearner(args..., strats...)
 end
-
-# add to an existing meta
-function make_learner(meta::MetaLearner, args...; kw...)
```
Comment (Evizero) [inline src/metalearner.jl]: Removing this looks like a breaking change. is it? Why did you remove this method?

**Review by Evizero — APPROVED** [review]
nice

File: src/metalearner.jl (hunk 5 above)
Comment (joshday) [reply src/metalearner.jl]: ...on accident
